codec-http2/src/main/java/io/netty/handler/codec/http2/ChildChannel.java

```diff
@@ -0,0 +1,275 @@
+        if (stackDepth < MAX_READER_STACK_DEPTH) {
+            threadLocals.setLocalChannelReaderStackDepth(stackDepth + 1);
+            try {
+                readTask.run();
+            } finally {
+                threadLocals.setLocalChannelReaderStackDepth(stackDepth);
+            }
+        } else {
+            eventLoop().execute(readTask);
+        }
+    }
+
+    @Override
+    protected final void doWrite(ChannelOutboundBuffer in) throws Exception {
+        if (closed) {
+            throw new ClosedChannelException();
+        }
+
+        EventExecutor preferredExecutor = doWritePreferredEventExecutor();
+
+        if (preferredExecutor.inEventLoop()) {
+            for (;;) {
+                Object msg = in.current();
+                if (msg == null) {
+                    break;
+                }
+                try {
+                    doWrite(msg);
+                } catch (Throwable t) {
+                    pipeline().fireExceptionCaught(t);
+                }
+                in.remove();
+            }
+            doWriteComplete();
+        } else {
+            // Use a copy because the original msgs will be recycled by AbstractChannel.
+            final Object[] msgsCopy = new Object[in.size()];
+            for (int i = 0; i < msgsCopy.length; i ++) {
+                msgsCopy[i] = ReferenceCountUtil.retain(in.current());
+                in.remove();
+            }
+
+            preferredExecutor.execute(new Runnable() {
+                @Override
+                public void run() {
+                    for (Object msg : msgsCopy) {
+                        try {
+                            doWrite(msg);
+                        } catch (Throwable t) {
+                            pipeline().fireExceptionCaught(t);
+                        }
+                        ReferenceCountUtil.release(msg);
+                    }
+                    doWriteComplete();
+                }
+            });
+        }
+    }
+
+    protected abstract void doWrite(Object msg) throws Exception;
+
+    protected abstract void doWriteComplete();
+
+    protected abstract EventExecutor doWritePreferredEventExecutor();
+
+    protected abstract void channelReadOccurred(Object msg);
+
+    protected void fireChildRead(final Object msg) {
+        if (eventLoop().inEventLoop()) {
+            fireChildRead0(msg);
+        } else {
+            eventLoop().execute(new OneTimeTask() {
+                @Override
+                public void run() {
+                    fireChildRead0(msg);
+                }
+            });
+        }
+    }
+
+    private void fireChildRead0(Object msg) {
+        if (readInProgress) {
+            assert inboundBuffer.isEmpty();
+            pipeline().fireChannelRead(msg);
+        } else {
+            inboundBuffer.add(msg);
+        }
+    }
+
+    protected void fireChildReadComplete() {
+        if (eventLoop().inEventLoop()) {
+            fireChildReadCompleteTask.run();
```

Comment:
Can you provide a more concrete reason why this is a bad idea? Performance-wise it seems it would be similar to calling any non-static helper method, which we do frequently without being concerned with function call overhead.

With the current code it is _very_ easy to see the two code paths do the same thing. A static method is still not as readable because the Runnable is at the top of the file, and you have to know it calls the same function in the inEventLoop() case.